swarm/src/behaviour/peer_addresses.rs

```diff
@@ -0,0 +1,306 @@
+use crate::behaviour::FromSwarm;
+use crate::{DialError, DialFailure, NewExternalAddrOfPeer};
+
+use libp2p_core::Multiaddr;
+use libp2p_identity::PeerId;
+
+use lru::LruCache;
+
+use std::collections::HashSet;
+use std::num::NonZeroUsize;
+
+/// Struct for tracking peers' external addresses of the [`Swarm`](crate::Swarm).
+#[derive(Debug)]
+pub struct PeerAddresses(LruCache<PeerId, HashSet<Multiaddr>>);
+
+impl PeerAddresses {
+    /// Creates cache with capacity of `cache_size`.
+    pub fn new(cache_size: NonZeroUsize) -> Self {
+        Self(LruCache::new(cache_size))
+    }
+
+    /// Feed a [`FromSwarm`] event to this struct.
+    ///
+    /// Returns whether the event changed peer's known external addresses.
+    pub fn on_swarm_event(&mut self, event: &FromSwarm) -> bool {
+        match event {
+            FromSwarm::NewExternalAddrOfPeer(NewExternalAddrOfPeer { peer_id, addr }) => {
+                if let Some(peer_addrs) = self.0.get_mut(peer_id) {
+                    let addr = prepare_addr(peer_id, addr);
+                    peer_addrs.insert(addr)
+                } else {
+                    let addr = prepare_addr(peer_id, addr);
+                    self.put(*peer_id, std::iter::once(addr));
+                    true
+                }
+            }
+            FromSwarm::DialFailure(DialFailure {
+                peer_id: Some(peer_id),
+                error: DialError::NoAddresses,
+                ..
+            }) => self.0.pop(peer_id).is_some(),
+            _ => false,
+        }
+    }
+
+    /// Adds addresses to cache.
+    /// Appends addresses to the existing set if peer addresses already exist.
+    /// Creates a new cache entry for peer_id if no addresses are present.
+    /// Returns true if the newly added addresses were not previously in the cache.
+    pub fn put(&mut self, peer: PeerId, addresses: impl Iterator<Item = Multiaddr>) -> bool {
+        let addresses = addresses.filter_map(|a| a.with_p2p(peer).ok());
+        if let Some(cached) = self.0.get_mut(&peer) {
+            let mut inserted_any = false;
+            for addr in addresses {
+                if cached.insert(addr) {
+                    inserted_any = true;
+                }
+            }
+
+            inserted_any
```

Comment:
Wanted to do it with some iterator, but couldn't come up with a neat solution - thus the regular for loop :/